satellite/gracefulexit/endpoint.go

```diff
@@ -862,3 +686,216 @@ func (endpoint *Endpoint) updatePointer(ctx context.Context, originalPointer *pb
+		node, err := endpoint.overlaydb.UpdateExitStatus(ctx, request)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		err = endpoint.db.IncrementProgress(ctx, nodeID, 0, 0, 0)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+		// graceful exit initiation metrics
+		age := time.Now().UTC().Sub(node.CreatedAt.UTC())
+		mon.FloatVal("graceful_exit_init_node_age_seconds").Observe(age.Seconds())
+		mon.IntVal("graceful_exit_init_node_audit_success_count").Observe(node.Reputation.AuditSuccessCount)
+		mon.IntVal("graceful_exit_init_node_audit_total_count").Observe(node.Reputation.AuditCount)
+		mon.IntVal("graceful_exit_init_node_piece_count").Observe(node.PieceCount)
+
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
+	}
+
+	if exitStatus.ExitLoopCompletedAt == nil {
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
+	}
+
+	return nil, nil
+}
+
+func (endpoint *Endpoint) generateExitStatusRequest(ctx context.Context, progress *Progress) (*overlay.ExitStatusRequest, error) {
+	processed := progress.PiecesFailed + progress.PiecesTransferred
+
+	if processed > 0 {
+		mon.IntVal("graceful_exit_successful_pieces_transfer_ratio").Observe(progress.PiecesTransferred / processed)
+	}
+
+	exitStatusRequest := &overlay.ExitStatusRequest{
+		NodeID:         progress.NodeID,
+		ExitFinishedAt: time.Now().UTC(),
+	}
+	// check node's exiting progress to see if it has failed passed max failure threshold
+	if processed > 0 && float64(progress.PiecesFailed)/float64(processed)*100 >= float64(endpoint.config.OverallMaxFailuresPercentage) {
+		exitStatusRequest.ExitSuccess = false
+	} else {
+		exitStatusRequest.ExitSuccess = true
+	}
+
+	if exitStatusRequest.ExitSuccess {
+		mon.Meter("graceful_exit_success").Mark(1)
+	} else {
+		mon.Meter("graceful_exit_fail_max_failures_percentage").Mark(1)
+	}
+
+	return exitStatusRequest, nil
+
+}
+
+func (endpoint *Endpoint) validatePointer(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem) (*pb.RemotePiece, error) {
+	remote := pointer.GetRemote()
+	nodeID := incomplete.NodeID
+
+	pieces := remote.GetRemotePieces()
+	var nodePiece *pb.RemotePiece
+	for _, piece := range pieces {
+		if piece.NodeId == nodeID && piece.PieceNum == incomplete.PieceNum {
+			nodePiece = piece
+		}
+	}
+
+	if nodePiece == nil {
+		endpoint.log.Debug("piece no longer held by node", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+
+		err := endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+	}
+	return nodePiece, nil
+}
+
+func (endpoint *Endpoint) generateFindNodesRequest(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem) (*overlay.FindStorageNodesRequest, error) {
```

Comment:
I think this name is misleading since this function also handles the logic of removing a piece from the pointer and transfer queue if the redundancy is high enough. I think either the name should be changed or the logic related to the find nodes request should be separated from the other logic here.